File: frontend/src/services/api/company.ts
/**
 * Company API service.
 * All calls use the existing `request()` client which auto-attaches the Bearer token.
 *
 * Endpoints consumed:
 *   POST /auth/company-signup  — atomic registration (auth router)
 *   GET  /company/me           — fetch own company profile
 *   PATCH /company/me          — update own company profile
 */

import { request } from "./client";
import type { Company, CompanySession, CompanySignupPayload, CompanyUpdatePayload } from "@/types/company";

// ── Shape helpers (snake_case DB → camelCase TS) ───────────────────────

function normalizeCompany(raw: Record<string, unknown>): Company {
  return {
    id: raw.id as string,
    ownerId: raw.owner_id as string,
    name: raw.name as string,
    industry: raw.industry as string,
    size: raw.size as string,
    hiringDomains: (raw.hiring_domains as string[]) ?? [],
    website: (raw.website as string | null) ?? null,
    logoUrl: (raw.logo_url as string | null) ?? null,
    isVerified: Boolean(raw.is_verified),
    createdAt: raw.created_at as string,
  };
}

// ── Service ────────────────────────────────────────────────────────────

export const companyService = {
  /**
   * Atomic company + HR account creation.
   * Calls POST /auth/company-signup (no auth required — creates the user).
   */
  signup: (payload: CompanySignupPayload): Promise<CompanySession> =>
    request<Record<string, unknown>>(
      "/auth/company-signup",
      {
        method: "POST",
        body: {
          email: payload.email,
          password: payload.password,
          first_name: payload.firstName,
          last_name: payload.lastName,
          company_name: payload.companyName,
          industry: payload.industry,
          size: payload.size,
          hiring_domains: payload.hiringDomains,
        },
      },
    ).then((raw) => ({
      user: raw.user as CompanySession["user"],
      token: raw.token as string,
      expiresAt: raw.expiresAt as string,
      company: raw.company ? normalizeCompany(raw.company as Record<string, unknown>) : null,
    })),

  /**
   * Fetch the authenticated HR user's company profile.
   * Requires a valid Bearer token (company role).
   */
  getMe: (): Promise<Company> =>
    request<Record<string, unknown>>("/company/me").then(normalizeCompany),

  /**
   * Partially update the company profile (website, logo, domains, etc.).
   * Requires a valid Bearer token (company role).
   */
  updateMe: (payload: CompanyUpdatePayload): Promise<Company> =>
    request<Record<string, unknown>>(
      "/company/me",
      {
        method: "PATCH",
        body: {
          ...(payload.name !== undefined && { name: payload.name }),
          ...(payload.industry !== undefined && { industry: payload.industry }),
          ...(payload.size !== undefined && { size: payload.size }),
          ...(payload.hiringDomains !== undefined && { hiring_domains: payload.hiringDomains }),
          ...(payload.website !== undefined && { website: payload.website }),
          ...(payload.logoUrl !== undefined && { logo_url: payload.logoUrl }),
        },
      },
    ).then(normalizeCompany),
};

File: backend/app/routers/company/company.py
"""Company router — GET /company/me · PATCH /company/me

All routes are protected: require a valid Bearer token AND role == "company".
Uses the service-role db_client, so RLS is bypassed on the backend.
"""

import logging
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from postgrest.exceptions import APIError

from ...deps import require_company_role
from ...schemas import CompanyOut, CompanyUpdateIn
from ...crud import get_company_by_owner_id, update_company, upsert_company

log = logging.getLogger(__name__)
router = APIRouter(prefix="/company", tags=["company"])


# ── Helpers ────────────────────────────────────────────────────────────

def map_company(row: dict) -> dict:
    """Convert a DB companies row to the CompanyOut shape."""
    return {
        "id": row["id"],
        "owner_id": row["owner_id"],
        "name": row["name"],
        "industry": row["industry"],
        "size": row["size"],
        "hiring_domains": row.get("hiring_domains") or [],
        "website": row.get("website"),
        "logo_url": row.get("logo_url"),
        "is_verified": bool(row.get("is_verified", False)),
        "created_at": str(row.get("created_at", "")),
    }


# ── GET /company/me ────────────────────────────────────────────────────

@router.get("/me", response_model=CompanyOut)
def get_my_company(
    current_user: dict = Depends(require_company_role),
) -> dict:
    """Return the company profile for the authenticated HR user."""
    owner_id = current_user["id"]
    try:
        company = get_company_by_owner_id(owner_id)
    except APIError as e:
        log.error("DB error fetching company for owner %s: %s", owner_id, e)
        raise HTTPException(status_code=500, detail="Failed to fetch company profile.")

    if not company:
        raise HTTPException(
            status_code=404,
            detail="Company profile not found. Please complete onboarding.",
        )
    return map_company(company)


# ── PATCH /company/me ──────────────────────────────────────────────────

@router.patch("/me", response_model=CompanyOut)
def update_my_company(
    payload: CompanyUpdateIn,
    current_user: dict = Depends(require_company_role),
) -> dict:
    """Update the caller's company profile — or create it on first run.

    Accounts that reached the company portal via Google / generic signup
    (not the atomic /auth/company-signup form) have a profile row but no
    companies row yet; company onboarding is where that row gets created,
    so this endpoint upserts. Inserts require name + industry + size
    (all NOT NULL in the schema)."""
    owner_id = current_user["id"]

    # Build the DB write dict — only include provided fields
    update_data: dict = {}
    if payload.name is not None:
        update_data["name"] = payload.name
    if payload.industry is not None:
        update_data["industry"] = payload.industry
    if payload.size is not None:
        update_data["size"] = payload.size
    if payload.hiring_domains is not None:
        update_data["hiring_domains"] = payload.hiring_domains
    if payload.website is not None:
        update_data["website"] = payload.website
    if payload.logo_url is not None:
        update_data["logo_url"] = payload.logo_url

    existing = get_company_by_owner_id(owner_id)

    if existing:
        try:
            company = update_company(owner_id, update_data)
        except APIError as e:
            log.error("DB error updating company for owner %s: %s", owner_id, e)
            raise HTTPException(status_code=400, detail=f"Update failed: {e.message}")
    else:
        missing = [f for f in ("name", "industry", "size") if not update_data.get(f)]
        if missing:
            raise HTTPException(
                status_code=422,
                detail=f"Cannot create company profile — missing required field(s): {', '.join(missing)}.",
            )
        try:
            company = upsert_company(owner_id, update_data)
        except APIError as e:
            log.error("DB error creating company for owner %s: %s", owner_id, e)
            raise HTTPException(status_code=400, detail=f"Create failed: {e.message}")

    if not company:
        raise HTTPException(status_code=404, detail="Company profile not found.")

    return map_company(company)
